🚀 2단계 - 리팩터링(메뉴)

writer0713 opened the pull request:
안녕하세요, 리뷰어님.
2단계 진행 중간에 pr 드립니다.

현재까지 아래 사항들 진행했습니다.
- 이전에 주신 피드백 수정
- `products` : 
  - `application`, `domain`, `infra` 구현 및 테스트 추가
- `menus` : 
  - `domain` 구현
  - `application` 에 `service` 구현 중 막힘 (아래 궁금한 부분들 좀 더 공부하고 추가로 진행하려 합니다)
    - `menu` 를 저장할때 이미 생성된 `menuGroup` 을 같이 저장해야하는지?
      - `menu` 에서 `menuGroupId` 외에는 딱히 `menuGroup` 의 속성을 사용하고 있지 않아서 굳이 직접 참조가 필요할까 싶네요.
    - `menu` 를 저장할때 `menuProduct` 의 `productId` 로 `Product` 가 존재하는지 확인
      - 이 로직이 어디에 위치해야 하는지? (domain Service??)

리뷰 잘 부탁드립니다. 감사합니다 :)

writer0713 committed:
refactor: Product Entity / VO 수정
- private 기본 생성자 추가
- fields final 제거

writer0713 committed:
test: 사용하지 않는 메서드 제거

writer0713 committed:
test: Product 의 가격(Price) 에 대한 비교를 객체비교로 수정
- Price 객체 내의 int 형 price 를 꺼내서 비교하지 않고 Price 객체 자체를 비교하도록 수정

writer0713 committed:
test: @NullAndEmptySource 어노테이션 사용
- null 과 emptyString 모두를 테스트 하기 위해서 @NullAndEmptySource 사용하도록 수정

writer0713 committed:
refactor: name 에 대한 validation 을 DisplayedName 으로 이동

writer0713 committed:
feat: MenuGroup 도메인 추가

writer0713 committed:
test: MenuGroup 테스트 추가

writer0713 committed:
refactor: MenuGroupName 에 hashCode 와 equals 메서드 오버라이드

writer0713 committed:
feat: TobeMenuProduct 추가
- embeded Price(가격) 추가
- embeded Quantity(수량) 추가

writer0713 committed:
test: TobeMenuProductTest 추가

writer0713 committed:
refactor: @Column 어노테이션을 @Embeddable 클래스 내로 이동

writer0713 committed:
feat: TobeMenu entity 추가

writer0713 committed:
refactor: 테스트 클래스 패키지 수정

writer0713 committed:
refactor: Profanities interface 를 shared.domain 패키지 하위로 이동
- Product 와 Menu 에서 공통으로 쓰는 공유커널로 변경

writer0713 committed:
feat: TobeMenu 추가
- TobeMenuProducts VO 추가

writer0713 committed:
feat: TobeMenuProduct 에 getTotalPrice 메서드 구현
- 메서드 구현을 위해 Price 와 Quantity 에 getter 추가

writer0713 committed:
test: TobeMenuTest 추가

writer0713 committed:
test: TobeMenuProduct getTotalPrice() 테스트 추가

writer0713 committed:
test: TobeMenuProductsTest 추가

writer0713 committed:
feat: TobeMenu 도메인 정책 추가
- 메뉴에 속한 상품 금액의 합은 메뉴의 가격보다 크거나 같아야 한다.
- 관련 테스트 추가

writer0713 committed:
feat: 메뉴 가격 변경 가능하도록 기능 추가
- 관련 테스트 추가 (메뉴의 가격은 0원 이상이어야 한다)

writer0713 committed:
feat: TobeMenu 에 display / hide 메서드 추가
- 메뉴 노출 / 숨김 기능 추가

writer0713 committed:
feat: TobeProduct id getter 추가

writer0713 committed:
refactor: fk 중복 에러로 인해 fk명 변경

writer0713 committed:
feat: TobeProduct 의 repository 와 jpaRepository 추가

writer0713 committed:
feat: TobeProductService 추가

writer0713 committed:
test: TobeProductServiceTest 추가

writer0713 committed:
feat: 테스트용 InMemoryTobeProductRepository 추가

writer0713 committed:
test: JpaTobeProductRepositoryTest 추가

writer0713 committed:
feat: TobeProductService 구현

writer0713 committed:
test: TobeProductServiceTest 추가

writer0713 committed:
chore: menus tobe.domain 패키지 이동
- menus.domain.tobe.domain => menus.tobe.domain 으로 이동

writer0713 committed:
feat: menuGroup 관련 repository 추가

writer0713 committed:
feat: menu 관련 repository 추가

writer0713 committed:
test: JpaTobeMenuRepositoryTest 추가

writer0713 committed:
test: TobeMenuGroupServiceTest 추가

writer0713 committed:
test: TobeMenuGroupServiceTest 에 @Transactional 추가

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/application/TobeMenuGroupService.kt:
```diff
@@ -0,0 +1,25 @@
+@Service
+class TobeMenuGroupService(
+    private val tobeMenuGroupRepository: TobeMenuGroupRepository,
+) {
+    @Transactional
+    fun create(request: TobeMenuGroup): TobeMenuGroup {
```
별도로 request, response Dto가 나와야 하지 않을까요?

writer0713 replied:
@hyunssooo 
네네, 안그래도 `TobeMenu` 쪽 application service 를 만들다보니 `dto` 가 필요할것 같아서 우선 `TobeMenu` 쪽만 `dto` 생성해서 작업하고 있었습니다 :)

이 부분도 dto 추가하도록 하겠습니다.

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/application/TobeMenuGroupService.kt:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import java.util.*
+
+@Service
+class TobeMenuGroupService(
```
서비스만 코틀린이군요? 🙂

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/DisplayedName.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Column(name = "name", nullable = false)
+    private String name;
+
+    private DisplayedName() {
+    }
```
protected 혹은 public으로 선언되어야 합니다. 보통은 protected 로 많이 선언해요. 
몇몇 기본 생성자가 생략되어 있는 것 같은데요. 이 부분은 따로 챙겨보면 좋을 것 같습니다. 
https://wbluke.tistory.com/6

writer0713 replied:
proxy 객체가 기본 생성자를 호출하는 시점에서 문제가 되는군요..!
감사합니다 :) 
수정하겠습니답

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenu.java:
```diff
@@ -0,0 +1,69 @@
+        this.menuProducts = TobeMenuProducts.of(tobeMenuProducts);
+        if (price > this.menuProducts.getTotalPrice()) {
+            throw new IllegalStateException("메뉴에 속한 상품 금액의 합은 메뉴의 가격보다 크거나 같아야 한다.");
+        }
```
잘 만들어둔 price 객체에게 메세지를 보내 보는 것은 어떨까요? 

```java
public class Money implements Comparable<Money> {

  private final BigDecimal value;
  // 생성자 메서드 equals & hashCode 생략

  public boolean isGreaterThanOrEqual(Money other) {
    return this.value.compareTo(other.value) >= 0;
  }

  public boolean isLessThanOrEqual(Money other) {
    return this.value.compareTo(other.value) <= 0;
  }
}
```

아래와 같은 메서드를 제공할 수 있지 않을까요?

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menuproduct/TobeMenuProduct.java:
```diff
@@ -0,0 +1,37 @@
+
+    public int getTotalPrice() {
+        return price.getPrice() * quantity.getQuantity();
+    }
```
`Price#times` 혹은 `Price#multiply` 같은 메서드를 제공할 수 도 있을 것 같아요.

writer0713 committed:
feat: default constructor 추가

writer0713 committed:
feat: DefaultProfanities 추가

writer0713 committed:
feat: id getter 추가

writer0713 committed:
refactor: of 메서드 대신 public constructor 로 대체

writer0713 committed:
feat: TobeMenuGroup 을 포함하는 생성자 추가

writer0713 committed:
feat: CreateMenuRequest dto 추가

writer0713 committed:
test: JpaTobeMenuRepositoryTest 추가

writer0713 committed:
feat: TobeMenuService 추가
- 메뉴 등록 기능 추가 (validation 아직 없음)

writer0713 committed:
test: TobeMenuServiceTest 추가

writer0713 committed:
feat: MenuProductRequest Dto 추가
- CreateMenuRequest 의 menuProducts 에 맞는 MenuProductRequest Dto 추가

writer0713 committed:
test: TobeMenuServiceTest 테스트 케이스 추가

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/application/TobeMenuService.kt:
```diff
@@ -0,0 +1,32 @@
+) {
+    @Transactional
+    fun create(request: CreateMenuRequest): TobeMenu? {
+        val menuGroup =
+            tobeMenuGroupRepository.findById(request.groupId)
+                .orElseThrow { NoSuchElementException() }
```
이번 미션과 관련은 없어 중요하지 않지만 서비스를 코틀린으로 작성한다면 Repository도 코틀린으로 작성해도 괜찮을 것 같아요. 

https://ttl-blog.tistory.com/840

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenu.java:
```diff
@@ -0,0 +1,83 @@
+    @Embedded
+    private Price price;
+
+    @ManyToOne(optional = false)
+    @JoinColumn(
+            name = "menu_group_id",
+            columnDefinition = "binary(16)",
+            foreignKey = @ForeignKey(name = "fk_menu_to_menu_group_2")
+    )
+    private TobeMenuGroup menuGroup;
```
> menu 를 저장할때 이미 생성된 menuGroup 을 같이 저장해야하는지?

두 엔티티는 같은 aggregate에 속하나요? 어떻게 생각하시나요?

writer0713 replied:
오늘 계속 생각을 해봤는데 정리된 생각은 아래와 같습니다.
- `menu` 가 `menuGroup` 의 속성을 사용하는 요구사항은 없다.
- `menu` 생성시 `menuGroup` 의 `id` 로 `이미 등록된 menuGroup` 인지 확인하는데에만 사용된다.
- `menu` 와 `menuGroup` 의 라이프사이클이 다르다.
  - `menuGroup` 단독으로 먼저 생성
  - `menu` 가 이후에 추가될때 `menuGroup` 의 존재만 확인

따라서, 같은 aggregate 가 아니라고 생각합니다.
`menu` 에서 직접참조 (`@ManyToOne`) 필요 없이 `groupId` 만 가지고 있으면 될것 같습니다.

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenuProducts.java:
```diff
@@ -0,0 +1,39 @@
+package kitchenpos.menus.tobe.domain.menu;
+
+import jakarta.persistence.*;
+import kitchenpos.menus.tobe.domain.menuproduct.TobeMenuProduct;
```
menu와 menuproduct의 패키지는 나눠지는데요. 어떤 기준일까요?

writer0713 replied:
보기 편하게 정리하다보니 그렇게 나눴는데 `aggregate` 기준으로 나눈다면 아래와 같이 나누는게 더 적절할것 같습니다.
- menu
  - menu + menuproduct
- menugroup
  - menugroup

hyunssooo commented on src/main/java/kitchenpos/products/tobe/application/TobeProductService.kt:
```diff
@@ -0,0 +1,38 @@
+class TobeProductService(
+    private val tobeProductRepository: TobeProductRepository,
+) {
+    @Transactional
+    fun create(request: TobeProduct): TobeProduct {
+        return tobeProductRepository.save(request)
+    }
+
+    @Transactional
+    fun changePrice(
+        productId: UUID,
+        request: TobeProduct,
+    ): TobeProduct {
```
request, response dto를 별도로 정의하는 것은 어떨까요?

writer0713 replied:
@hyunssooo 
네네, 처음에 dto 없이 entity 로만 작업을 하다가 필요에 따라 우선 `request dto` 만 생성을 해서 작업중이었습니다.
현재 `domain service` 를 어떻게 적용해야할지 고민중이다보니 `response dto` 는 아직 신경을 못쓰고 있었습니다 :)

hyunssooo commented on src/main/java/kitchenpos/products/tobe/application/TobeProductService.kt:
```diff
@@ -0,0 +1,38 @@
+        return tobeProductRepository.save(product)
+    }
+
+    @Transactional(readOnly = true)
+    fun findAll(): MutableList<TobeProduct> {
+        return tobeProductRepository.findAll()
+    }
```
MutableList 보다 immutable list를 반환하는게 좋을 것 같아요. :)

hyunssooo commented on src/main/java/kitchenpos/products/tobe/domain/Price.java:
```diff
@@ -9,7 +9,10 @@
     @Column(name = "price", nullable = false)
-    private final int price;
+    private int price;
+
+    private Price() {
+    }
```
price는 모든 BC에서 공통으로 사용될 것 같은데 각각 선언해서 사용해야하나요? 어떻게 생각하실까요?

writer0713 replied:
요 부분도 고민하던 부분중 하나였는데요.
만약 현재 프로젝트를 한 팀에서 관리한다면 공통으로 사용해도 될것 같습니다.
반대로 각 bc 를 각각의 팀에서 관리한다고 보면 각 bc 에서 각자의 price 를 갖는게 나을거라 생각이 듭니다.

현재 작업은 한 팀에서 현재 프로젝트 (모든 bc) 를 관리한다고 생각하고 시작했으니 price 의 경우 `공유 커널` 로 사용해도 될것 같습니다 :)

혹시 잘못 생각한 부분 있다면 코멘트 부탁드립니다.

hyunssooo commented on src/main/java/kitchenpos/shared/domain/DefaultProfanities.kt:
```diff
@@ -0,0 +1,12 @@
+package kitchenpos.shared.domain
+
+class DefaultProfanities : Profanities {
+    private val profanities =
+        setOf<String>(
+            "비속어",
+        )
```
```suggestion
    private val profanities =
        setOf(
            "비속어",
        )
```

hyunssooo commented on src/test/java/kitchenpos/menus/tobe/application/TobeMenuGroupServiceTest.kt:
```diff
@@ -0,0 +1,54 @@
+
+@Transactional
+@SpringBootTest
+class TobeMenuGroupServiceTest
+    @Autowired
+    constructor(
+        private val sut: TobeMenuGroupService,
+    ) {
```
```suggestion
class TobeMenuGroupServiceTest(
  @Autowired
  private val sut: TobeMenuGroupService,
) {
```

`@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)` 을 사용해
혹은 생성자 주입을 사용해도 괜찮을 것 같아요. 

```suggestion
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
class TobeMenuGroupServiceTest(
  private val sut: TobeMenuGroupService,
) {
```

writer0713 committed:
chore: TODO 코멘트를 메서드 상단 코멘트로 이동

hyunssooo commented on src/test/java/kitchenpos/menus/tobe/domain/menu/TobeMenuProductsTest.java:
```diff
@@ -0,0 +1,38 @@
+
+class TobeMenuProductsTest {
+
+    @DisplayName("")
+    @Test
+    void case_1() {
+        // given
+        var menuProduct1 = createTobeMenuProduct(10_000, 2);
+        var menuProduct2 = createTobeMenuProduct(13_000, 1);
+        var menuProduct3 = createTobeMenuProduct(12_000, 1);
+        var menuProducts = createTobeMenuProducts(menuProduct1, menuProduct2, menuProduct3);
+
+        // when
+        int totalPrice = menuProducts.getTotalPrice();
+
+        // then
+        assertThat(totalPrice).isEqualTo(45_000);
+
+    }
```
테스트의 의도를 명확히 나타내면 좋을 것 같아요.

writer0713 replied:
에고, `DisplayName` 을 깜빡했네요 :)

hyunssooo commented on src/test/java/kitchenpos/menus/tobe/domain/menu/TobeMenuTest.java:
```diff
@@ -0,0 +1,103 @@
+                .isThrownBy(() -> menu.changePrice(-1));
+    }
+
+    private TobeMenu createTobeMenu(int price) {
+        var name = "후라이드";
+        var profanities = new FakeProfanities("비속어");
+        var menuGroupId = UUID.randomUUID();
+        var menuProducts = List.of(new TobeMenuProduct(1, 10_000, UUID.randomUUID()));
+
+        return new TobeMenu(name, price, profanities, menuGroupId, menuProducts);
+    }
+
+    private TobeMenu createTobeMenu(String name) {
+        var price = 10_000;
+        var profanities = new FakeProfanities("비속어");
+        var menuGroupId = UUID.randomUUID();
+        var menuProducts = List.of(new TobeMenuProduct(1, 10_000, UUID.randomUUID()));
+
+        return new TobeMenu(name, price, profanities, menuGroupId, menuProducts);
+    }
+
+    private TobeMenu createTobeMenu(List<TobeMenuProduct> menuProducts) {
+        var name = "후라이드";
+        var price = 10_000;
+        var profanities = new FakeProfanities("비속어");
+        var menuGroupId = UUID.randomUUID();
+
+        return new TobeMenu(name, price, profanities, menuGroupId, menuProducts);
+    }
```
각 메서드보다 메서드들을 사용하는 부분에서 생성자를 사용하는게 더 나을 것 같기도 하네요. 🤔
의견 주시면 좋을 것 같아요.

writer0713 replied:
name, profanities, menuGroupId, menuProducts 네 가지 속성을 매번 건드려야 하다보니 중복이 많다고 생각되어 세가지 케이스로 fixture(?) 를 만들었습니다 :) 

장점으로는 중복으로 타이핑을 안해도 된다는점이 있을것 같고,
단점으로는 파라미터로 넘기는 값(ex. price) 를 제외하면 다른 속성이 어떤값을 가지는지 바로 파악이 어렵다는 점이 있을듯 합니다.

hyunssooo commented on src/test/java/kitchenpos/menus/tobe/domain/menu/TobeMenuTest.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+class TobeMenuTest {
+
```
성공하는 테스트도 함께 작성되면 좋을 것 같아요.

hyunssooo reviewed (changes requested):
안녕하세요.
몇 가지 코멘트 남겼습니다.
함께 고민해 보면 좋을 것 같아요. 🙂

writer0713 committed:
feat: existsById 메서드 추가

writer0713 committed:
refactor: TobeMenu 리팩토링
- menuGroup 직접참조 제거
- menuGroupId 로 간접참조하고 validation 시에만 활용
- menuGroup 포함한 생성자 삭제

hyunssooo commented:
> menu 를 저장할때 menuProduct 의 productId 로 Product 가 존재하는지 확인
이 로직이 어디에 위치해야 하는지? (domain Service??)

저라면 menuProduct 생성 전 으로 위치할 것 같습니다. 
가볍게 validation정도만 하고 바로 저장해도 괜찮을 것 같아요.

writer0713 commented:
> > menu 를 저장할때 menuProduct 의 productId 로 Product 가 존재하는지 확인
> > 이 로직이 어디에 위치해야 하는지? (domain Service??)
> 
> 저라면 menuProduct 생성 전 으로 위치할 것 같습니다. 가볍게 validation정도만 하고 바로 저장해도 괜찮을 것 같아요.

코멘트 주신 부분을 아래와 같이 이해해도 될까요?
- menuService (application service) 에서 MenuRepository 와 ProductRepository 를 가지고 validation 을 진행한다.

hyunssooo commented:
앗 아닙니다. 이 부분은 ProductRepository에 대한 의존은 menuService에서 의존하지 않을 것 같아요. 추상화를 통해 간접적으로 의존할 것 같아요. BC가 별도의 서비스로 분리된다면 해당 부분만 DB -> api 조회로 변경될 것 같습니다.

writer0713 committed:
refactor: menus 패키지 수정
- menu 와 menugroup 패키지만 남김
- menu 에 'menu' 와 'menuProduct' 도메인 같이 위치

writer0713 committed:
refactor: product 관련 entity / vo 기본 생성자 modifier protected 로 수정

writer0713 committed:
refactor: menugroup 관련 entity / vo 기본 생성자 modifier protected 로 수정

writer0713 committed:
refactor: menu, menuProduct 관련 entity / vo 기본 생성자 modifier protected 로 수정

writer0713 committed:
refactor: Price 리팩터링
- menu 와 product 에서 공통의 Price 객체 사용하도록 수정
- Price Comparable 구현
- Price 에 isGreaterThan 메서드 추가

writer0713 committed:
refactor: Price 에 multiply 메서드 추가
- TobeMenuProduct.getTotalPrice() 에서 price.multiply 사용하도록 수정

writer0713 committed:
refactor: TobeProductService 의 findAll 메서드 리턴타입 수정
- MutableList 대신 List 로 수정

writer0713 committed:
refactor: setOf 의 타입추론 사용하도록 수정

writer0713 committed:
test: @TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL) 사용하도록 수정

writer0713 committed:
test: DisplayName 명시

writer0713 committed:
refactor: TobeMenu 생성자 signature 순서 변경 및 displayed 추가

writer0713 committed:
feat: TobeMenuFactory 추가
- dto => TobeMenu 생성하는 부분과 validation 을 TobeMenuFactory 에 위임

writer0713 committed:
feat: TobeProductClient 추가
- 구현체 DefaultTobeProductClient 추가

writer0713 committed:
test: DefaultTobeProductClientTest 추가

writer0713 committed:
refactor: TobeMenuFactory 에 validateExistsProduct 메서드 구현
- TobeProductClient 를 사용하여 validation 진행

writer0713 committed:
test: TobeMenuServiceTest 테스트 수정 및 추가
- 1 개 이상의 등록된 상품으로 메뉴를 등록할 수 있다.

writer0713 committed:
feat: 메뉴 목록 조회 구현
- test 추가

writer0713 committed:
feat: 메뉴 숨김 / 노출 / 가격변경 구현
- 테스트 추가

writer0713 committed:
feat: changePriceRequest 추가

writer0713 committed:
refactor: menu 관련 responseDto 추가

writer0713 committed:
refactor: product 관련 request / response dto 추가

writer0713 committed:
refactor: menuGroup 관련 request / response dto 추가 및 수정

writer0713 commented:
@hyunssooo 
안녕하세요, 리뷰어님.
말씀주신 피드백 포함하여 아래 내용들 수정했습니다.
- dto 추가 
- default constructor protected 로 변경
- Price 객체 shared 패키지로 이동
- Price 객체 Comparable 구현
- Price 객체 isGreaterThan 메서드 추가
- Price 객체 multply 메서드 추가
- profanities setOf 타입추론 사용하도록 수정 
- 테스트 @TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL) 사용하도록 수정 
- TobeMenuFactory 추가
- TobeProductClient 인터페이스 추가 
- DefaultTobeProductClient 구현체 추가 (TobeProductRepository 사용)
- 테스트 수정 / 추가

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenu.java:
```diff
@@ -0,0 +1,90 @@
+    @Embedded
+    private TobeMenuProducts menuProducts;
+
+    @Transient
+    private UUID menuGroupId;
```
간접 참조를 하는 건 좋은데 `@Transient` 으로 선언돼도 괜찮나요?

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenuProduct.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Embedded
+    private Price price;
+
+    @Transient
+    private UUID productId;
+
```
여기도 확인이 필요할 것 같아요.

hyunssooo commented on src/main/java/kitchenpos/shared/domain/Price.java:
```diff
@@ -38,4 +43,17 @@ public boolean equals(Object o) {
+        return this.price - other.price;
+    }
+
+    public boolean isGreaterThan(int price) {
+        return this.price - price > 0;
+    }
+
+    public int multiply(Quantity quantity) {
+        return this.price * quantity.getQuantity();
+    }
```
`Price#isGreaterThan` 같은 경우는 파라미터로 다른 price 객체를 받는 것이 좀 더 자연스럽지 않나요? 
```
`가격`은 다른 `가격`과 더 큰지 비교할 수 있다.
```

```java
  public boolean isGreaterThan(Price other) {
    return this.price > other.price;
  }
```

`Price#multiply` 같은 경우는 price 객체를 리턴 하는 것이 조금 더 자연스러운 것 같아요.

사소하지만 일반적인 메서드 순서도 확인해 보시면 좋을 것 같아요. 일반 메서드들이 Override 된 메서드들 보다 상위에 위치합니다. :)

hyunssooo replied:
테스트도 작성해 보면 좋을 것 같아요.

hyunssooo commented on src/main/java/kitchenpos/menus/tobe/domain/menu/TobeMenuProduct.java:
```diff
@@ -0,0 +1,47 @@
+        return productId;
+    }
+
+    public TobeMenuProductResponse toDto() {
+        return new TobeMenuProductResponse(seq, quantity.getQuantity(), price.getPrice(), productId);
+    }
```
고민해 볼만한 포인트인 거 같아요. domain 모델이 dto에 의존하고 있는데요. 이 경우는 괜찮을까요? 어떻게 생각하시나요?

hyunssooo reviewed (approved):
안녕하세요. 
몇 가지 가벼운 코멘트 남겼습니다. 
다음 step 진행하시면서 확인 부탁드려요!
